### PR TITLE
Add option to use scheduled sampling in CopyNet

### PR DESCRIPTION
**Pull request by JohnGiorgi**
This PR adds the ability to use scheduled sampling in `CopyNetSeq2Seq` by supplying an argument for `scheduled_sampling_ratio` that's greater than zero. It is essentially a copy/paste from `SimpleSeq2Seq`.

This helps reduce the differences in the `SimpleSeq2Seq` and `CopyNetSeq2Seq` model arguments. It is also backwards compatible with a default value of 0 (no scheduled sampling i.e. teacher forcing).

### CODE REVIEW

**Review by epwalsh — CHANGES_REQUESTED** [review]
Thanks @JohnGiorgi! I think this is a good addition. I just have a couple of suggestions:
- I think we should have a test
- I think the default for `scheduled_sampling_ratio` should be `None` so that when it is `None` we don't have to call `torch.rand()`. That way there is no performance penalty for this feature.

**Comment by JohnGiorgi** [comment]
@epwalsh Awesome, thanks for the feedback. 

- Updated the `if` statement so that if `scheduled_sampling_ratio` is falsey, `torch.rand` is not called. You could set the default to `None`, but this makes the default value for this parameter different than `simple_seq2seq`.
- Added a simple test that the model can train with `0 < scheduled_sampling_ratio < 1`. I wasn't really sure how to write a more significant test due to the randomness. Let me know if you have a different test in mind!

I could also update `simple_seq2seq` so that `torch.rand` is not called when `scheduled_sampling_ratio` is falsey, to save a tiny bit of performance there!

File: allennlp_models/generation/models/copynet_seq2seq.py [hunk 1]
```diff
@@ -514,13 +524,33 @@ def _forward_loss(
...
-            copied = ((input_choices == self._oov_index) & (target_to_source.sum(-1) > 0)).long()
+            if (
+                self.training
+                and self._scheduled_sampling_ratio
+                and torch.rand(1).item() < self._scheduled_sampling_ratio
+            ):
+                # Use gold tokens at test time and at a rate of 1 - _scheduled_sampling_ratio
+                # during training.
+                # shape: (batch_size,)
+                input_choices = last_predictions
```
Comment (JohnGiorgi) [inline allennlp_models/generation/models/copynet_seq2seq.py:543]: Ah, I realized this implementation doesn't work, because `last_predictions` is never updated. I would have had to take the index of the token with the highest probability for this timestep under the model. Something like:

```python
last_predictions = torch.max(torch.cat((copy_scores, generation_scores), -1), -1)
```

@epwalsh does this make sense?

File: tests/generation/models/copynet_test.py [hunk 2]
```diff
@@ -44,6 +44,13 @@ def test_model_can_train_with_amp(self):
             overrides="{'trainer.use_amp':true,'trainer.cuda_device':0}",
         )
 
+    def test_model_can_train_with_scheduled_sampling_ratio(self):
+        train_model_from_file(
+            self.param_file,
+            self.TEST_DIR,
+            overrides="{'model.scheduled_sampling_ratio':0.5}",
+        )
```
Comment (epwalsh) [inline tests/generation/models/copynet_test.py:52]: > Added a simple test that the model can train with 0 < scheduled_sampling_ratio < 1. I wasn't really sure how to write a more significant test due to the randomness. Let me know if you have a different test in mind!

I think this is totally fine for a test. I just want to have coverage of the new lines of code to ensure that the model trains without error.

File: allennlp_models/generation/models/copynet_seq2seq.py (hunk 1 above)
Comment (JohnGiorgi) [reply allennlp_models/generation/models/copynet_seq2seq.py:543]: Gotcha. Could I just return [`log_probs`](https://github.com/allenai/allennlp-models/blob/86444374c487f1d6991cd83b7a3da80b026effc2/allennlp_models/generation/models/copynet_seq2seq.py#L430) from `_get_ll_contrib()`? Its computed like: `log_probs = util.masked_log_softmax(all_scores, mask)`.

Comment (epwalsh) [reply allennlp_models/generation/models/copynet_seq2seq.py:543]: Yes, good point.

Comment (JohnGiorgi) [reply allennlp_models/generation/models/copynet_seq2seq.py:543]: Awesome, just pushed that change.

File: tests/generation/models/simple_seq2seq_test.py [hunk 3]
```diff
@@ -46,6 +46,13 @@ def test_model_can_train_with_amp(self):
             overrides="{'trainer.use_amp':true,'trainer.cuda_device':0}",
         )
 
+    def test_model_can_train_with_scheduled_sampling_ratio(self):
+        train_model_from_file(
+            self.param_file,
+            self.TEST_DIR,
+            overrides="{'model.scheduled_sampling_ratio':0.5}",
+        )
+
```
Comment (JohnGiorgi) [inline tests/generation/models/simple_seq2seq_test.py:55]: @epwalsh Added the sample test for scheduled sampling to `simple_seq2seq`.

File: allennlp_models/generation/models/simple_seq2seq.py [hunk 4]
```diff
@@ -367,7 +367,11 @@ def _forward_loop(
         step_logits: List[torch.Tensor] = []
         step_predictions: List[torch.Tensor] = []
         for timestep in range(num_decoding_steps):
-            if self.training and torch.rand(1).item() < self._scheduled_sampling_ratio:
+            if (
+                self.training
+                and self._scheduled_sampling_ratio > 0.0
+                and torch.rand(1).item() < self._scheduled_sampling_ratio
+            ):
```
Comment (JohnGiorgi) [inline allennlp_models/generation/models/simple_seq2seq.py:374]: @epwalsh Added a similar condition to `simple_seq2seq` to avoid the call to `torch.rand` when `_scheduled_sampling_ratio` is `0.0`.

**Review by epwalsh — COMMENTED** [review]
This LGTM! Can you just update the CHANGELOG? Then I think this is good to go.

**Comment by JohnGiorgi** [comment]
Cool! Changelog updated 👍

**Review by epwalsh — APPROVED** [review]
